### PR TITLE
Fix PR9494 (one case of recursive module initialisation)

### PR DESCRIPTION
**Pull request by mshinwell**
As per the discussion on #9494.  (This PR is against 4.11.)

### CODE REVIEW

File: stdlib/camlinternalMod.ml [hunk 1]
```diff
@@ -54,9 +54,10 @@ let rec update_mod shape o n =
       (* The optimisation below is invalid on bytecode since
          the RESTART instruction checks the length of closures.
          See PR#4008 *)
```
Comment (xavierleroy) [inline stdlib/camlinternalMod.ml]: The comment needs updating:
In bytecode, the RESTART instruction checks the size of closures.
Hence, the optimized case overwrite o n is valid only if o and n have the same size.
(See PR#4008.)
In native-code, the size of closures does not matter, so overwriting is possible
as soon as the size of n is less than that of o.
```

**Review by xavierleroy — APPROVED** [review]
LGTM.  One suggestion below.

File: stdlib/camlinternalMod.ml (hunk 1 above)
Comment (mshinwell) [reply stdlib/camlinternalMod.ml]: Thanks for the suggestion, I've committed a slightly modified version of this.

**Comment by xavierleroy** [comment]
Committed to 4.11 + cherry-picked to trunk df20ccf83

**Pull request merged**